File: HW/HW12/minCostSupplyWater.py
class UnionFind:
    def __init__(self,size):
        self.parent = list(range(size))
        self.rank = [0] * size

    def find(self, p):
        #Path compression
        if self.parent[p] != p:
            se
            lf.parent[p] = self.find(self.parent[p]) 
        return self.parent[p]

    def unify(self, p, q):
        rootP = self.find(p)
        rootQ = self.find(q)
        if rootP != rootQ: #Lets unify by rank
            if self.rank[rootP] > self.rank[rootQ]:
                self.parent[rootQ] = rootP
            elif self.rank[rootP] < self.rank[rootQ]:
                self.parent[rootP] = rootQ
            else:
                self.parent[rootQ] = rootP
                self.rank[rootP] += 1

class Edge:
    def __init__(self, src, dest, cost):
        self.cost = cost
        self.src = src
        self.dest = dest

    #compareTo in python
    def __lt__(self, object):
        return self.cost < object.cost



def minCostToSupplyWater(n: int, wells, pipes) -> int:
    edgeList = []
    for i in range(1, n+1):
        edgeList.append(Edge(0,i, wells[i-1]))
    for pipe in pipes:
        edgeList.append(Edge(pipe[0], pipe[1], pipe[2]))

    edgeList.sort()
    uf = UnionFind(n+1)

    cost = 0
    for edge in edgeList:
        rootSrc = uf.find(edge.src)
        rootDest = uf.find(edge.dest)
        if (rootSrc == rootDest):
            continue
        uf.unify(rootSrc, rootDest)
        cost += edge.cost
        n -=1
        if n==0:
            break

    return cost

n1 = 3
wells1 = [1, 2, 3]
pipes1 = []
print(minCostToSupplyWater(n1, wells1, pipes1))

n2 = 1
wells2 = [10]
pipes2 = []
print(minCostToSupplyWater(n2, wells2, pipes2))

File: HW/HW12/minCostSupplyWater.java
package HW.HW12;


import java.util.*;

public class minCostSupplyWater {

    public static class Edge implements Comparable<Edge>
    {
        int src;
        int dest; 
        int weight;
        public Edge(int src, int dest, int weight)
        {
            this.src = src;
            this.dest = dest;
            this.weight = weight;
        }

        @Override
        public int compareTo(Edge o)
        {
            return this.weight - o.weight;
        }
    }

    public static class UnionFind
    {
        int[] parent;
        int[] size;

        public UnionFind(int n)
        {
            //Set every node as each other parent
            parent = new int [n + 1];
            for (int i = 1; i < parent.length; i++)
            {
                parent[i] = i;
            }
            size = new int[n+1];
        }

        public int find(int n)
        {
            //Finds the root of n
            int root = n;
            while (root != parent[root])
            {
                root = parent[root];
            }
            //Compress the path to root, essentially
            //points all nodes in the path from n to root
            //to the root, which compresses the nodes
            while (n != root)
            {
                int currNext = parent[n];
                parent[n] = root;
                n = currNext;
            }
            return root;
        }

        public void unify(int x, int y)
        {
            int rootx = find(x);
            int rooty = find(y);
            //Already in the same set
            if (rootx == rooty)
            {
                return;
            }
            if (size[rootx] > size[rooty])
            {
                //We add size and change the pointers of parent of smaller  size
                size[rootx] += size[rooty];
                parent[rooty] = x;
            }
            else
            {
                size[rooty] += size[rootx];
                parent[rootx] = y;
            }
        }

    }
    
    public static int minCostToSupplyWater(int n, int[] wells, int[][] pipes) 
    {
        // Your code here
        List<Edge> edgeList = new ArrayList<>();
        //Think of a well as a singular node, connect all nodes to that singular node with a weight
        for (int i = 1; i <= wells.length; i++)
        {
            edgeList.add(new Edge(0, i, wells[i-1]));
        }

        //Add the weights of all pipes
        for (int [] pipe: pipes)
        {
            edgeList.add(new Edge(pipe[0],pipe[1], pipe[2]));
        }

        //Sort the weights
        Collections.sort(edgeList);
        UnionFind union = new UnionFind(n);
        int total = 0;
        for (Edge edge: edgeList)
        {
            int root1 = union.find(edge.src);
            int root2 = union.find(edge.dest);
            if (root1 == root2)
            {
                //They are already connected
                continue;
            }
            //Aren't connected we unify them toghether
            total += edge.weight;
            union.unify(root1, root2);
            //Important to update the number of houses because 
            //there can be multiple edges pointing to one house, so we can cut loop short if we found the best path
            n--;
            if (n == 0)
            {
                return total;
            }
        }
        return total;
     }

     public static void main(String [] args)
     {
        int n1 = 3;
        int[] wells1 = {1, 2, 3};
        int[][] pipes1 = {};
        System.out.println(minCostToSupplyWater(n1, wells1, pipes1));  // Test Case 1 = 6
        int n2 = 1;
        int[] wells2 = {10};
        int[][] pipes2 = {};
        System.out.println(minCostToSupplyWater(n2, wells2, pipes2));  // Test Case 2 = 10
     }
}
